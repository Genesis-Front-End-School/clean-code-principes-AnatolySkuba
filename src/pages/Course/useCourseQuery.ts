import { useQuery } from "react-query";
import { getCourse } from "../../services/API";
import { ROUTER_KEYS } from "../../utils/constants";

export const useCourseQuery = (courseId: string) => {
    const { data, isLoading, isError } = useQuery(
        [`${ROUTER_KEYS.COURSES}`, { courseId }],
        getCourse
    );
    return { data, isLoading, isError };
};
