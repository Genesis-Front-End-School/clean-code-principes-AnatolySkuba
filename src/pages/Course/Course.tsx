import React, { useEffect, useState } from "react";
import {
    FcAcceptDatabase,
    FcPositiveDynamic,
    FcStart,
    FcUndo,
    FcVoicePresentation,
} from "react-icons/fc";
import { SlLock } from "react-icons/sl";
import { useNavigate, useParams } from "react-router-dom";
import { toast } from "react-toastify";
import Hls from "hls.js";

import { STORAGE_KEYS } from "../../utils/constants";
import { Lesson } from "../../utils/types";
import { useCourseQuery } from "./useCourseQuery";

export const Course = (): JSX.Element => {
    const [videoEl, setVideoEl] = useState<HTMLVideoElement | null>(null);
    const [isPip, setIsPip] = useState("");
    const navigate = useNavigate();
    const { courseId } = useParams<{ courseId: string }>();
    const { data, isLoading, isError } = useCourseQuery(courseId!);

    useEffect(() => {
        if (courseId) {
            const video = document.getElementById(courseId) as HTMLVideoElement;
            !videoEl && setVideoEl(video);
        }
    }, [courseId, videoEl, isLoading]);

    if (isLoading) {
        return <p data-testid="course-loading">Loading...</p>;
    }

    if (isError) {
        return <div data-testid="course-error">Error!</div>;
    }

    const { id, title, description, lessons, rating, meta } = data?.data;

    if (
        videoEl &&
        Hls.isSupported() &&
        meta.courseVideoPreview?.link &&
        localStorage.getItem(STORAGE_KEYS.TOKEN)
    ) {
        const savedTime = localStorage.getItem(`currentCourseTime ${id}`);
        const hls = new Hls({
            xhrSetup: (xhr) => {
                xhr.responseType = "json";
                xhr.setRequestHeader(
                    "Authorization",
                    `Bearer ${localStorage.getItem(STORAGE_KEYS.TOKEN)}`
                );
            },
        });
        hls.loadSource(meta.courseVideoPreview?.link);
        // hls.loadSource("https://test-streams.mux.dev/x36xhzz/x36xhzz.m3u8");
        hls.attachMedia(videoEl);

        savedTime && hls.startLoad(Number(savedTime));
        setTimeout(function () {
            if (savedTime) videoEl.currentTime = Number(savedTime);
            videoEl.play();
        }, 150);
        videoEl.onpause = () => {
            localStorage.setItem(`currentCourseTime ${id}`, videoEl?.currentTime.toString());
        };
    }

    const pip = async (id: string) => {
        const video = document.getElementById(id) as HTMLVideoElement;

        const eventListenerKeydown = (event: KeyboardEvent) => {
            if (event.ctrlKey) {
                event.code === "Comma" &&
                    (video.playbackRate =
                        video.playbackRate >= 0.5 ? video.playbackRate - 0.25 : 0.25);
                event.code === "Period" &&
                    (video.playbackRate =
                        video.playbackRate <= 3.75 ? video.playbackRate + 0.25 : 4);
            }
        };
        window.addEventListener("keydown", eventListenerKeydown, false);

        if (
            Hls.isSupported() &&
            meta.courseVideoPreview?.link &&
            localStorage.getItem(STORAGE_KEYS.TOKEN)
        ) {
            const savedTime = localStorage.getItem(`currentLessonTime ${id}`);
            const hls = new Hls({
                xhrSetup: (xhr) => {
                    xhr.responseType = "json";
                    xhr.setRequestHeader(
                        "Authorization",
                        `Bearer ${localStorage.getItem(STORAGE_KEYS.TOKEN)}`
                    );
                },
            });
            hls.loadSource(meta.courseVideoPreview?.link);
            // hls.loadSource("https://test-streams.mux.dev/x36xhzz/x36xhzz.m3u8");
            hls.attachMedia(video);
            savedTime && hls.startLoad(Number(savedTime));
            setTimeout(function () {
                if (savedTime) video.currentTime = Number(savedTime);
                video.play();
            }, 150);
            video.onpause = () => {
                localStorage.setItem(`currentLessonTime ${id}`, video?.currentTime.toString());
            };
        }

        video.addEventListener("enterpictureinpicture", () => {
            setIsPip(id);
        });
        video.addEventListener("leavepictureinpicture", () => {
            window.removeEventListener("keydown", eventListenerKeydown, false);
            setIsPip("");
            localStorage.setItem(`currentLessonTime ${id}`, video?.currentTime.toString());
        });

        video.onloadedmetadata = async () => {
            try {
                if (document.pictureInPictureElement) {
                    localStorage.setItem(`currentLessonTime ${id}`, video?.currentTime.toString());
                    await document.exitPictureInPicture();
                }
                video?.requestPictureInPicture();
            } catch (err) {
                toast.error("Video failed to enter/leave Picture-in-Picture mode.");
                console.log("Video failed to enter/leave Picture-in-Picture mode.", err);
            }
        };
    };

    return (
        <div className="p-3" data-testid="course-page">
            <FcUndo className="cursor-pointer" onClick={() => navigate(-1)} role="link" />
            <div className="max-w-160 mx-auto">
                <video
                    id={id}
                    className="border h-full w-full rounded-md object-cover"
                    loop
                    muted
                    controls
                    data-testid="video-player"
                />
            </div>
            <div className="mt-1">
                <p className="font-medium">{title}</p>
                <p className="text-sm text-gray-500">{description}</p>

                <div className="flex items-center justify-around text-xs">
                    <div className="mt-1.5">
                        <p className="text-gray-500">Rating</p>
                        <p className="flex font-medium gap-1">
                            <FcPositiveDynamic size="16" />
                            {rating}
                        </p>
                    </div>

                    <div className="mt-1.5">
                        <div className="flex gap-2">
                            <p className="text-gray-500">Skills</p>
                            <FcVoicePresentation size="16" />
                        </div>
                        <ul>
                            {meta.skills ? (
                                meta.skills?.map((skill: string, index: number) => (
                                    <li key={index}>
                                        <p className="w-50 font-medium">{skill}</p>
                                    </li>
                                ))
                            ) : (
                                <li>
                                    <p>?</p>
                                </li>
                            )}
                        </ul>
                    </div>
                </div>

                <div className="mt-2">
                    <div className="flex gap-2 items-center">
                        <p className="font-medium">Lessons</p>
                        <FcAcceptDatabase size="16" />
                    </div>

                    <ul>
                        {lessons?.map((lesson: Lesson) => (
                            <li key={lesson.id} className="flex items-center gap-3">
                                <video className="hidden" id={lesson.id} loop muted></video>
                                <p
                                    className="cursor-pointer text-blue-800 hover:scale-105 hover:ml-2"
                                    onClick={() =>
                                        lesson.status === "locked"
                                            ? toast.error("Oops, the lesson is locked!")
                                            : pip(lesson.id)
                                    }
                                    role="button"
                                >
                                    {lesson.title}
                                </p>
                                {lesson.status === "locked" && <SlLock size="10" color="red" />}
                                {lesson.id === isPip && (
                                    <>
                                        <FcStart className="animate-pulse" />
                                        <p>
                                            Press Ctrl + &#62; to increase speed, press Ctrl + &lt;
                                            to slow down playback
                                        </p>
                                    </>
                                )}
                            </li>
                        ))}
                    </ul>
                </div>
            </div>
        </div>
    );
};
