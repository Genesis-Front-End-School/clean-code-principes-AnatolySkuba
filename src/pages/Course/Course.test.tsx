import { screen } from "@testing-library/react";

import { Course } from "./Course";
import { useCourseQuery } from "./useCourseQuery";
import { renderWithRouter } from "../../tests/helpers/renderWithRouter";

const mockedUseCourseQuery = useCourseQuery as jest.Mock;
jest.mock("./useCourseQuery");

const response = {
    data: {
        containsLockedLessons: true,
        description: "Find the inner strength to overcome procrastination and reach your goals.",
        duration: 509,
        id: "3b77ceb6-fb43-4cf5-a25b-8fe9222a0714",
        launchDate: "2023-03-06T16:25:24.000Z",
        lessons: [
            {
                duration: 229,
                id: "b9ad7391-0f0b-4fe1-b919-6525d76ad3c4",
                link: "https://wisey.app/videos/the-power-of-self-discipline-how-to-stay-on-track/lesson-1/AppleHLS1/lesson-1.m3u8",
                meta: null,
                order: 1,
                previewImageLink:
                    "https://wisey.app/assets/images/web/lessons-covers/the-power-of-self-discipline-how-to-stay-on-track/lesson-1",
                status: "unlocked",
                title: "Why short-term gains aren’t worth it",
                type: "video",
            },
        ],
        meta: {
            slug: "the-power-of-self-discipline-how-to-stay-on-track",
            skills: ["Increasing self-awareness"],
            courseVideoPreview: {
                link: "https://wisey.app/videos/the-power-of-self-discipl…w-to-stay-on-track/preview/AppleHLS1/preview.m3u8",
                duration: 19,
                previewImageLink:
                    "https://wisey.app/assets/images/web/course-covers/the-power-of-self-discipline-how-to-stay-on-track/preview",
            },
        },
        previewImageLink:
            "https://wisey.app/assets/images/web/course-covers/the-power-of-self-discipline-how-to-stay-on-track",
        rating: 3.5,
        status: "launched",
        tags: ["productivity"],
        title: "The Power of Self-Discipline: How to Stay on Track",
    },
};

describe("Course component", () => {
    afterEach(jest.clearAllMocks);

    test("Displays the loading view", () => {
        mockedUseCourseQuery.mockImplementation(() => ({
            isLoading: true,
        }));
        renderWithRouter(<Course />);
        expect(screen.getByTestId("course-loading")).toBeInTheDocument();
        expect(screen.getByText(/Loading.../i)).toBeVisible();
    });

    test("Displays the error message", () => {
        mockedUseCourseQuery.mockImplementation(() => ({
            isError: true,
        }));
        renderWithRouter(<Course />);
        expect(screen.getByTestId("course-error")).toBeInTheDocument();
        expect(screen.getByText(/Error!/i)).toBeVisible();
    });

    test("Render Course correctly", () => {
        mockedUseCourseQuery.mockImplementation(() => ({
            data: response,
        }));
        renderWithRouter(<Course />);
        expect(mockedUseCourseQuery).toBeCalledTimes(1);
        expect(screen.getByTestId("course-page")).toBeInTheDocument();
        expect(screen.getByTestId("video-player")).toBeInTheDocument();
        expect(screen.getByText("Lessons")).toBeInTheDocument();
        expect(screen.getByText("Rating")).toBeInTheDocument();
        expect(screen.getByText("Skills")).toBeInTheDocument();
    });

    test("The video tag should have loop and controls attributes", () => {
        mockedUseCourseQuery.mockImplementation(() => ({
            data: response,
        }));
        renderWithRouter(<Course />);
        expect(screen.getByTestId("video-player")).toHaveAttribute("loop");
        expect(screen.getByTestId("video-player")).toHaveAttribute("controls");
    });

    test("Course snapshot", () => {
        mockedUseCourseQuery.mockImplementation(() => ({
            data: response,
        }));
        renderWithRouter(<Course />);
        expect(screen.getByTestId("course-page")).toMatchSnapshot();
    });
});
