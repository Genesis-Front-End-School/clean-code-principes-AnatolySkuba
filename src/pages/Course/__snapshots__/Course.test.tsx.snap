// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Course component Course snapshot 1`] = `
<div
  class="p-3"
  data-testid="course-page"
>
  <svg
    class="cursor-pointer"
    enable-background="new 0 0 48 48"
    fill="currentColor"
    height="1em"
    role="link"
    stroke="currentColor"
    stroke-width="0"
    version="1"
    viewBox="0 0 48 48"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      fill="#00BCD4"
    >
      <polygon
        points="5,18 19,6.3 19,29.7"
      />
      <path
        d="M28,14H16v8h12c2.8,0,5,2.2,5,5s-2.2,5-5,5h-3v8h3c7.2,0,13-5.8,13-13S35.2,14,28,14z"
      />
    </g>
  </svg>
  <div
    class="max-w-160 mx-auto"
  >
    <video
      class="border h-full w-full rounded-md object-cover"
      controls=""
      data-testid="video-player"
      id="3b77ceb6-fb43-4cf5-a25b-8fe9222a0714"
      loop=""
    />
  </div>
  <div
    class="mt-1"
  >
    <p
      class="font-medium"
    >
      The Power of Self-Discipline: How to Stay on Track
    </p>
    <p
      class="text-sm text-gray-500"
    >
      Find the inner strength to overcome procrastination and reach your goals.
    </p>
    <div
      class="flex items-center justify-around text-xs"
    >
      <div
        class="mt-1.5"
      >
        <p
          class="text-gray-500"
        >
          Rating
        </p>
        <p
          class="flex font-medium gap-1"
        >
          <svg
            enable-background="new 0 0 48 48"
            fill="currentColor"
            height="16"
            stroke="currentColor"
            stroke-width="0"
            version="1"
            viewBox="0 0 48 48"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="#00BCD4"
            >
              <rect
                height="20"
                width="10"
                x="19"
                y="22"
              />
              <rect
                height="34"
                width="10"
                x="32"
                y="8"
              />
              <rect
                height="12"
                width="10"
                x="6"
                y="30"
              />
            </g>
            <g
              fill="#3F51B5"
            >
              <polygon
                points="11,8 21,18 21,8"
              />
              <rect
                height="14.1"
                transform="matrix(-.707 -.707 .707 -.707 10.879 36.506)"
                width="4"
                x="11"
                y="8.9"
              />
            </g>
          </svg>
          3.5
        </p>
      </div>
      <div
        class="mt-1.5"
      >
        <div
          class="flex gap-2"
        >
          <p
            class="text-gray-500"
          >
            Skills
          </p>
          <svg
            enable-background="new 0 0 48 48"
            fill="currentColor"
            height="16"
            stroke="currentColor"
            stroke-width="0"
            version="1"
            viewBox="0 0 48 48"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M40,22h-8l-4,4V12c0-2.2,1.8-4,4-4h8c2.2,0,4,1.8,4,4v6C44,20.2,42.2,22,40,22z"
              fill="#2196F3"
            />
            <circle
              cx="17"
              cy="19"
              fill="#FFA726"
              r="8"
            />
            <path
              d="M30,36.7c0,0-3.6-6.7-13-6.7S4,36.7,4,36.7V40h26V36.7z"
              fill="#607D8B"
            />
          </svg>
        </div>
        <ul>
          <li>
            <p
              class="w-50 font-medium"
            >
              Increasing self-awareness
            </p>
          </li>
        </ul>
      </div>
    </div>
    <div
      class="mt-2"
    >
      <div
        class="flex gap-2 items-center"
      >
        <p
          class="font-medium"
        >
          Lessons
        </p>
        <svg
          enable-background="new 0 0 48 48"
          fill="currentColor"
          height="16"
          stroke="currentColor"
          stroke-width="0"
          version="1"
          viewBox="0 0 48 48"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="#D1C4E9"
          >
            <path
              d="M38,7H10C8.9,7,8,7.9,8,9v6c0,1.1,0.9,2,2,2h28c1.1,0,2-0.9,2-2V9C40,7.9,39.1,7,38,7z"
            />
            <path
              d="M38,19H10c-1.1,0-2,0.9-2,2v6c0,1.1,0.9,2,2,2h28c1.1,0,2-0.9,2-2v-6C40,19.9,39.1,19,38,19z"
            />
            <path
              d="M38,31H10c-1.1,0-2,0.9-2,2v6c0,1.1,0.9,2,2,2h28c1.1,0,2-0.9,2-2v-6C40,31.9,39.1,31,38,31z"
            />
          </g>
          <circle
            cx="38"
            cy="38"
            fill="#43A047"
            r="10"
          />
          <polygon
            fill="#DCEDC8"
            points="42.5,33.3 36.8,39 34.1,36.3 32,38.5 36.8,43.3 44.6,35.5"
          />
        </svg>
      </div>
      <ul>
        <li
          class="flex items-center gap-3"
        >
          <video
            class="hidden"
            id="b9ad7391-0f0b-4fe1-b919-6525d76ad3c4"
            loop=""
          />
          <p
            class="cursor-pointer text-blue-800 hover:scale-105 hover:ml-2"
            role="button"
          >
            Why short-term gains arenâ€™t worth it
          </p>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
